appsrc/dcfile/dcuidchg.man
.TH DCUIDCHG 1 "11 March 2007" "DICOM PS3" "DICOM PS3 - Change UIDs"
.SH NAME
dcuidchg \- ACR/NEMA DICOM PS3 ... DICOM PS3 - Change UIDs consistently
.SH SYNOPSIS
.HP 10
.B dcuidchg
" inputfile1 [ inputfile2 ... ]"
.B \-outdir " dirname"
[
.B \-v|verbose|vv|veryverbose
]
[
.B \-l
]
[
.B \-f|-filelist " filename"
]
[
.B \-descriptor
]
[
.B \-map " filename"
]
.so man1/optin.so
.so man1/optout.so
.SH DESCRIPTION
.LP
.B dcuidchg
reads all the named dicom input files (and/or the files listed in the specified
file) and copies them to the specified directory, changing all the instance-related UIDs
in a consistent manner throughout the set of files.
.LP
Fixed UIDs such as SOP Classes and Transfer Syntaxes are not affected.
.LP
The UIDs obtained by two separate invocations of this utility will be different, so
the action needs to be performed on a set of files simultaneously in order
to maintain consistency across studies, series, frames of reference and
in explicit references to other instances.
.LP
Note that the basename of the original file is used for the output file, and is
not changed, so if the basename was derived from the SOP Instance UID, as is common
practice, then that name would need to be changed subsequently if the UID replacement
was being performed to prevent identity leakage.
.LP
.SH OPTIONS
The verbose output goes to standard error.
.LP
The basic switches are described in dcintro(1). Options specific to this program are:
.TP
.B \-outdir dirname
.RS
The directory, which must already exist, to which the copied files are written, using
the basename of the original filename.
.RE
.TP
.B \-v|verbose
.RS
Display attributes as they are read and written.
.RE
.TP
.B \-l
.RS
List filenames as they are read and written.
.RE
.TP
.B \-f|filelist filename
.RS
A file containing a list of DICOM input files (one filename per line).
.RE
.TP
.B \-descriptor
.RS
Add (or extend) DeidentificationMethod attribute with description of UID removal.
.RE
.TP
.B \-map filename
.RS
A file to which is written the map of original to replacement UID values.
.RE
.SH ENVIRONMENT
.LP
\ 
.SH EXAMPLES
.LP
.RE
\ 
.RE
% ls -1 IMAGES/[0-9]*
.RE
IMAGES/IM0001
.RE
IMAGES/IM0002
.RE
% mkdir /tmp/IMAGES
.RE
% dcuidchg IMAGES/[0-9]* -outdir /tmp/NEWIMAGES
.RE
% ls -1 /tmp/NEWIMAGES
.RE
/tmp/NEWIMAGES/IM0001
.RE
/tmp/NEWIMAGES/IM0002
.SH FILES
.LP
\ 
.SH SEE ALSO
.BR dcintro(1), 
.BR dcdtchg(1)
.SH AUTHOR
Copyright (C) 1993-2010. David A. Clunie DBA PixelMed Publishing. All rights reserved.
.SH BUGS
.LP
There is a fixed length of buffer used for reading each line from the list of filenames file.
.LP


CHANGES
111219:	libsrc/src/dctool/attrmxrd.cc: check for HighBit being more than one less than BitsStored and correct BitsStored accordingly (otherwise will create zero pixel values on copy) (000408)

111213:	appsrc/dcfile/dcuncat.cc: add option to replace instance number starting with a specific number to allow scripts to extend a series with successive invocations

111212:	libsrc/standard/condn.tpl/module,strval/base.tpl,dx.tpl: improve animal versus human conditions to check SRT homo sapiens, and make view position and body part examined human versus animal specific

111212:	libsrc/standard/elmdict/other.tpl: add more Sound private elements

111117:	libsrc/standard/strval/base.tpl: add CSD for 99NCIAIM

111109:	appsrc/dcfile/dcuncat.cc: add options to replace instance number, frame increment pointer and frame increment pointer vectors, as well as add number of frames during unenhance based on SOP class

111106:	libsrc/standard/elmdict/other.tpl: add Vepro private elements

111031:	libsrc/standard/elmdict/other.tpl: add CTP private elements

111030:	libsrc/standard/elmdict/other.tpl: add Segami private elements

111029:	libsrc/standard/elmdict/dicom3.tpl: add (restore from trial) Observation UID (per CP 1147)

111028:	libsrc/standard/elmdict/other.tpl: add NeuroQuant private elements

111024:	libsrc/standard/elmdict/other.tpl: update fuji private attributes with descriptions

110918:	libsrc/standard/elmdict/diconde.tpl: update with new or modified data elements with cleanup from Patrick

110918:	libsrc/standard/elmdict/siemens.tpl: update some private elements based on Artis conformance statement

110916:	libsrc/standard/elmdict/gems.tpl: resolve inconsistencies in GE PET conformance statement VRs SS/SL/FL for private elements by choosing those most commonly seen in the field (still some inconsistencies between models)

110904:	libsrc/standard/elmdict/camtron.tpl: add private data elements encountered in DICOMDIR, as well as QCA private data elements from GEMS/Camtronics CRS conformance statement

110901:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Sup 152 FT ophthalmic thickness map (dictionary and SOP class only; not IOD yet)

110822:	libsrc/standard/elmdict/dicom3.tpl: update data dictionary (only) for CPs for FT Aug 2011 WG 6 meeting

110714:	libsrc/standard/elmdict/dicom3.tpl: remove 'unified' from names of various attributes (per Sup 96 FT)

110716:	libsrc/standard/strval/base.tpl: add 99PMP CSD

110714:	libsrc/standard/elmdict/dicom3.tpl: exposure index etc are DS not FD (CP 1024)

110709:	libsrc/standard/elmdict/dicom3.tpl: change (0018,9430) from US to FL per Sup 139

110612:	libsrc/standard/elmdict/diconde.tpl: update for DICONDE eddy current standard

110612:	libsrc/standard/elmdict/dicondep.tpl: correct case of DICONDE DX (Dx) private owner for E2699

110609:	libsrc/standard/elmdict/dicom3.tpl: add Sup 151 FT data elements

110608:	libsrc/standard/elmdict/elscint.tpl: update Elscint private elements

110531:	libsrc/standard/elmdict/other.tpl: update Hitachi CT private elements

110531:	libsrc/standard/elmdict/siemens.tpl: update mammo private elements

110529:	libsrc/standard/condn.tpl,iodcomp/dx.tpl,module/base.tpl,dx.tpl,xaxrf.tpl,strval/dx.tpl,xaxrf.tpl: add breast tomosynthesis IOD, refactoring 3D XRay accordingly

110529:	libsrc/standard/elmdict/other.tpl: more private data elements for Hologic 

110528:	appsrc/dcfile/dciodvfy.cc: do not complain about SCOORD without child IMAGE when SELECTED FROM is by reference (000406)

110527:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: check that references to instances in content tree are in top level evidence sequences in SR (000405)

110526:	libsrc/standard/elmdict/gems.tpl: add names for CT iterative reconstruction private data elements

110522:	libsrc/src/dctool/attrothr.cc: do not throw assertion if computed length is equal (which may occur for single bit images)

110522:	libsrc/standard/iodcomp/base.tpl,libsrc/standard/condn.tpl: permit FoR Module in Segementation (and do CP to standard to allow may be present otherwise)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect SCOORD without child IMAGE (000403)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect quotes in code meaning of measurement units (000404)

110522:	libsrc/src/dctool/attrtypv.cc,locale/mesgtext.tpl: permit PN with multiple carets in multiple component groups (000402)

110521:	libsrc/standard/elmdict/acuson.tpl,agfa.tpl,isg.tpl,other.tpl,siemens.tpl,strval/base.tpl: more private elements and coding scheme designators

110517:	libsrc/standard/module/base.tpl: correct type (1->2) and number of items (1->0-n) in SourceImageSequence in DerivationImageMacro

110517:	libsrc/standard/elmdict/siemens.tpl: yet more syngo private elements

110516:	libsrc/standard/cond.tpl: no longer require version for UCUM codes (CP 1060)

110516:	libsrc/standard/strval/base.tpl: add 99_MI private coding scheme designator

110516:	libsrc/include/dctool/attr.h,attrseq.h,generic/listsimp.h,src/dctool/attrseq.cc,attrvrfy.cc,standard/condn.tpl,module/softcopy.tpl,sr.tpl,support/module.awk:
		though VM for sequences number of items was being specified in modules, it was not being applied;
		apply it (000256), as well as make it work (since number of items was not being computed correctly
		in SequenceAttribute::verifyVM()); also add support for conditional check of VM, for example for
		GraphicData depending on Grahic Type (000401)

110516:	libsrc/standard/elmdict/toshiba.tpl: update CT private elements from conformance statement for V4.70 or later version

110514:	libsrc/standard/elmdict/siemens.tpl: add more syngo private elements, and update guessed VRs of existing ones with observed explicit VRs

110511:	libsrc/standard/elmdict/gems.tpl,toshiba.tpl: add more encountered explicit VR private elements

110511:	libsrc/standard/elmdict/other.tpl: update/add McKesson private elements from conformance statement


appsrc/dcfile/dcuidchg.cc
#include <fstream.h>

#include "attrmxls.h"
#include "attrtype.h"
#include "attrseq.h"
#include "mesgtext.h"
#include "dcopt.h"
#include "elmconst.h"

#include "hash.h"
#include "uidgen.h"

class UIDEntryString {
	HashKeyString key;
	const char *value;
public:
	UIDEntryString(void)		{}
	UIDEntryString(const char *s,const char *v)
					{ key=HashKeyString(s); value=v; }
	UIDEntryString(UIDEntryString *e)
					{ key=e->key; value=e->value; }

	HashKeyString	getKey(void) const	{ return key; }
	const char *	getValue(void) const	{ return value; }

	bool operator==(UIDEntryString e)
					{ return key == e.getKey(); }
};

class UIDEntryStringList : public SimpleList<UIDEntryString>
{
public:
	~UIDEntryStringList() {}	// only because buggy g++ 2.7.0 freaks
};

class UIDEntryStringListIterator : public SimpleListIterator<UIDEntryString>
{
public:
	UIDEntryStringListIterator(void)
		: SimpleListIterator<UIDEntryString>() {}
	UIDEntryStringListIterator(UIDEntryStringList& list)
		: SimpleListIterator<UIDEntryString>(list) {}
};

class UIDHashMap : public OpenHashTable <UIDEntryString,
					HashKeyString,
					UIDEntryStringList,
					UIDEntryStringListIterator>
{
public:
	UIDHashMap(unsigned long size)
		: OpenHashTable<UIDEntryString,
				HashKeyString,
				UIDEntryStringList,
				UIDEntryStringListIterator>(size)
		{}

	// supply virtual functions for OpenHashTable ...

	unsigned long	hash(const HashKeyString &key,unsigned long size)
		{
			const unsigned nchars = 10;
			const unsigned shift  = 4;

			unsigned n=nchars;
			unsigned long value=0;
			const char *s=key.getString();
			while (n-- && *s) {
				value=(value<<shift)|*s++;
			}
			return value%size; 
		}

	HashKeyString key(const UIDEntryString &e)	{ return e.getKey(); }
};

static bool
isAttributeToBeReplaced(Attribute *a) {
		Tag tag = a->getTag();
		return tag != TagFromName(SOPClassUID)
			&& tag != TagFromName(AffectedSOPClassUID)
			&& tag != TagFromName(MediaStorageSOPClassUID)
			&& tag != TagFromName(OriginalSpecializedSOPClassUID)
			&& tag != TagFromName(ReferencedRelatedGeneralSOPClassUIDInFile)
			&& tag != TagFromName(ReferencedSOPClassUID)
			&& tag != TagFromName(ReferencedSOPClassUIDInFile)
			&& tag != TagFromName(RelatedGeneralSOPClassUID)
			&& tag != TagFromName(RequestedSOPClassUID)
			&& tag != TagFromName(RelatedGeneralSOPClassUID)
			&& tag != TagFromName(SOPClassesInStudy)
			&& tag != TagFromName(SOPClassesSupported)
			&& tag != TagFromName(TransferSyntaxUID)
			&& tag != TagFromName(EncryptedContentTransferSyntaxUID)
			&& tag != TagFromName(MACCalculationTransferSyntaxUID)
			&& tag != TagFromName(ReferencedTransferSyntaxUIDInFile)
			&& tag != TagFromName(CodingSchemeUID)
			&& tag != TagFromName(ContextGroupExtensionCreatorUID)
			&& tag != TagFromName(PrivateInformationCreatorUID)
			&& tag != TagFromName(PrivateRecordUID);
}

static bool
loopOverListsInSequences(Attribute *a,UIDHashMap *uidHashMap,const char *uidstamp,ofstream *mapfstr,TextOutputStream &log,
		bool (*func)(AttributeList &,UIDHashMap *,const char *uidstamp,ofstream *mapfstr,TextOutputStream &))
{
	bool succeeded=true;
	if (strcmp(a->getVR(),"SQ") == 0) {
		AttributeList **al;
		int n;
		if ((n=a->getLists(&al)) > 0) {
			int i;
			for (i=0; i<n; ++i) {
				if (!(*func)(*al[i],uidHashMap,uidstamp,mapfstr,log)) succeeded=false;
			}
			delete [] al;
		}
	}
	return succeeded;
}

class OurGeneratedUID : public GeneratedUID {
public:
	OurGeneratedUID
			(const char *s,unsigned counter)
		: GeneratedUID(s)
		{
			setUnsigned(counter);
		}
};

static int uidcounter = 0;

static bool
changeUIDs(AttributeList& list,UIDHashMap *uidHashMap,const char *uidstamp,ofstream *mapfstr,TextOutputStream &log)
{
	bool succeeded=true;
	AttributeListIterator listi(list);
	while (!listi) {
		Attribute *a=listi();
		Assert(a);
		if (!::loopOverListsInSequences(a,uidHashMap,uidstamp,mapfstr,log,&::changeUIDs))
			succeeded=false;
		const char *vr = a->getVR();
		if (vr && strcmp(vr,"UI") == 0) {
			int n = a->getVM();
			int i;
			for (int i=0; i<n; ++i) {
				char *originalUID;
				if (a->getValue(i,originalUID) && originalUID && strlen(originalUID) && isAttributeToBeReplaced(a)) {
//cerr << "Original UID = " << originalUID << endl;
					const char *replacementUID = NULL;
					UIDEntryString *entry = (*uidHashMap)[originalUID];
					if (entry) {
//cerr << "Found entry = " << entry->getKey().getString() << ", " << entry->getValue() << endl;
						replacementUID = entry->getValue();
					}
					else {
						replacementUID = StrDup(OurGeneratedUID(uidstamp,uidcounter++));
//cerr << "Add entry = " << originalUID << ", " << replacementUID << endl;
						(*uidHashMap)+=new UIDEntryString(StrDup(originalUID),replacementUID);
						if (mapfstr) {
							(*mapfstr) << "\"" << originalUID << "\" \"" << replacementUID << "\"" << endl;
						}
					}
//cerr << "Replacement UID = " << replacementUID << endl;
					a->setValue(i,replacementUID);
				}
			}
		}
		//	succeeded=false;
		++listi;
	}
	return succeeded;
}

static void addUIDChangedToDeidentificationMethod(AttributeList& list)
{
	const char *newvalue = "replace uids (dcuidchg)";
	Attribute *aDeidentificationMethod = list[TagFromName(DeidentificationMethod)];
	if (aDeidentificationMethod) {
		aDeidentificationMethod->addValue(newvalue);
	}
	else {
		aDeidentificationMethod = new LongStringAttribute(TagFromName(DeidentificationMethod),newvalue);
		list+=aDeidentificationMethod;
	}
}

static const char *
makeNewFileName(const char *outdir,const char *filename) {
	Assert(outdir);
	Assert(filename);
	size_t outdirLength = strlen(outdir);
	size_t filenameLength = strlen(filename);
	const char *filenameLastSeparator = strrchr(filename,'/');
	if (filenameLastSeparator != NULL) {
		filename = filenameLastSeparator+1;
		filenameLength = strlen(filename);
//cerr << "makeNewFileName(): found last component " << filename << endl;
	}
	size_t length = outdirLength + 1 + filenameLength;
	char *string=new char[length+1];
	strcpy(string,outdir);
	strcat(string,"/");
	strcat(string,filename);
	string[length]=0;
	return string;
}

static void handleOneFile(UIDHashMap *uidHashMap,const char *outdir,const char *filename,
			DicomInputOptions &dicom_input_options,
			DicomOutputOptions &dicom_output_options,
			bool listnames,bool updatedescriptor,bool success,bool verbose,
			ofstream *mapfstr,
			TextOutputStream &log) {

		Assert(filename);
		if (listnames) log << "Reading \"" << filename << "\"" << endl;

#ifdef USEBINARYFLAGFORINPUTOPENMODE
		ifstream *ifstr=new ifstream(filename,ios::in|ios::binary);
#else
		ifstream *ifstr=new ifstream(filename);
#endif
		if (!ifstr || !*ifstr || !ifstr->rdbuf()->is_open()) {
			cerr << AMsgDC(FileReadOpenFailed);
			if (filename) cerr <<" - \"" << filename << "\"";
			cerr << endl;
			success=false;
			return;
		}

		DicomInputStream din(*(istream *)ifstr,
			dicom_input_options.transfersyntaxuid,
			dicom_input_options.usemetaheader);

		ManagedAttributeList list;

		if (verbose) log << "******** While reading ... " << filename << " ... ********" << endl;
		list.read(din,&log,verbose,0xffffffff,true,dicom_input_options.uselengthtoend,dicom_input_options.ignoreoutofordertags,dicom_input_options.useUSVRForLUTDataIfNotExplicit);

		if (!list.good()) {
			log << list.errors()
			    << EMsgDC(DatasetReadFailed) << endl;
			success=false;
			return;
		}

		// do NOT close input file yet, else deferred pixel data read for write will fail
		
		changeUIDs(list,uidHashMap,dicom_output_options.stamp,mapfstr,log);

		if (updatedescriptor) {
			addUIDChangedToDeidentificationMethod(list);
		}

		const char *outfilename = makeNewFileName(outdir,filename);
		Assert(outfilename);
		if (listnames) log << "Writing \"" << outfilename << "\"" << endl;
		
#ifdef USEBINARYFLAGFOROUTPUTOPENMODE
		ofstream *ofstr=new ofstream(outfilename,
			ios::out|ios::trunc|ios::binary);
#else
		ofstream *ofstr=new ofstream(outfilename,ios::out|ios::trunc);
#endif
		if (!ofstr || !*ofstr || !ofstr->rdbuf()->is_open()) {
			cerr << AMsgDC(FileWriteOpenFailed);
			if (outfilename) cerr <<" - \"" << outfilename << "\"";
			cerr << endl;
			success=false;
			return;
		}
		
		DicomOutputStream dout(*(ostream *)ofstr,
			dicom_output_options.transfersyntaxuid,
			dicom_output_options.usemetaheader,
			dicom_output_options.useimplicitmetaheader,
			dicom_output_options.addtiff);

		if (!usualManagedAttributeListWrite(list,dout,
			dicom_output_options,log,verbose)) success=false;
		
		if (ofstr) {
			ofstr->close();
			delete ofstr;
		}
		if (outfilename) {
			delete outfilename;
		}
		// Now can close input file
		if (ifstr) {
			ifstr->close();
			delete ifstr;
		}
}

int
main(int argc, char *argv[])
{
	bool bad=false;

	GetNamedOptions 	options(argc,argv);
	DicomInputOptions 	dicom_input_options(options);
	DicomOutputOptions 	dicom_output_options(options);

	bool verbose=options.get("verbose") || options.get("v");
	bool listnames=options.get("l");
	bool updatedescriptor=options.get("descriptor");

	const char *outdir=0;
	bad = !options.get("outdir",outdir) || bad;

	dicom_input_options.done();
	Assert(!dicom_input_options.filename);

	dicom_output_options.done();

	const char *filelistfile=0;
	(void)(options.get("filelist",filelistfile) || options.get("f",filelistfile));

	const char *mapfile=0;
	(void)(options.get("map",mapfile));

	int numberofinputfiles=!options;

	const char **listoffilenames = new const char * [numberofinputfiles];
	const char **ptr = listoffilenames;
	const char *filename;

	while(!options && (filename=options())) {
		++options;
		*ptr++=filename;
	}

	options.done();

	DicomOutputOpenerFromOptions output_opener(
		options,dicom_output_options.filename,cout);

	cerr << dicom_input_options.errors();
	cerr << dicom_output_options.errors();
	cerr << options.errors();
	cerr << output_opener.errors();

	if (!dicom_input_options.good()
	 || !dicom_output_options.good()
	 || !options.good()
	 || !output_opener.good()
	 || !options
	 || bad) {
		cerr 	<< MMsgDC(Usage) << ": " << options.command()
			<< dicom_input_options.usage()
			<< dicom_output_options.usage()
			<< " -outdir dirname"
			<< " [-v|-verbose]"
			<< " [-l]"
			<< " [-f|-filelist filename]"
			<< " [-map filenametowriteuidmap]"
			<< " [-descriptor]"
			<< " " << MMsgDC(InputFile) << " ["<< MMsgDC(InputFile) << " ...]"
			<< " >" << MMsgDC(OutputFile)
			<< endl;
		exit(1);
	}

	if (!dicom_output_options.transfersyntaxuid)
		dicom_output_options.transfersyntaxuid=ExplicitVRLittleEndianTransferSyntaxUID;

	bool success=true;
	TextOutputStream log(cerr);

	UIDHashMap *uidHashMap = new UIDHashMap(1000);

	ofstream *mapfstr=NULL;
	if (mapfile) {
		mapfstr=new ofstream(mapfile,ios::out|ios::trunc);
		Assert(mapfstr);
	}
	
	int i;
	for (i=0; i < numberofinputfiles; ++i) {
		handleOneFile(uidHashMap,outdir,listoffilenames[i],dicom_input_options,dicom_output_options,listnames,updatedescriptor,success,verbose,mapfstr,log);
	}

	if (filelistfile) {
		ifstream *flfstr=new ifstream(filelistfile);
		if (!flfstr || !*flfstr || !flfstr->rdbuf()->is_open()) {
			cerr << AMsgDC(FileReadOpenFailed);
			if (filelistfile) cerr <<" - \"" << filename << "\"";
			cerr << endl;
			bad=true;
		}
		else {
			while (flfstr->peek() != EOF) {
				const int lineBufferSize=2048;
				char lineBuffer[lineBufferSize];
				flfstr->getline(lineBuffer,2048);
				if (strlen(lineBuffer))
					handleOneFile(uidHashMap,outdir,lineBuffer,dicom_input_options,dicom_output_options,listnames,updatedescriptor,success,verbose,mapfstr,log);
				// else skip blank lines
			}
		}
	}
	
	if (mapfstr) {
		mapfstr->close();
		delete mapfstr;
	}

	if (numberofinputfiles && listoffilenames) delete[] listoffilenames;
	return success ? 0 : 1;
}

